<a href="https://colab.research.google.com/github/yaesur/class/blob/master/%ED%86%B5%EA%B3%84%EB%A1%9C_%EC%9A%94%EC%95%BD%ED%95%98%EA%B8%B0.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import gdown

gdown.download('https://bit.ly/3736JW1', 'ns_book6.csv', quiet=False)

Downloading...
From: https://bit.ly/3736JW1
To: /content/ns_book6.csv
100%|██████████| 55.0M/55.0M [00:00<00:00, 66.2MB/s]


'ns_book6.csv'

In [3]:
import pandas as pd

ns_book6 = pd.read_csv('ns_book6.csv', low_memory=False)
ns_book6.head()

,번호,도서명,저자,출판사,발행년도,ISBN,세트 ISBN,부가기호,권,주제분류번호,도서권수,대출건수,등록일자
0,1,인공지능과 흙,김동훈 지음,민음사,2021.0,9788937444319,NaN,NaN,NaN,NaN,1,0,2021-03-19
1,2,가짜 행복 권하는 사회,김태형 지음,갈매나무,2021.0,9791190123969,NaN,NaN,NaN,NaN,1,0,2021-03-19
2,3,나도 한 문장 잘 쓰면 바랄 게 없겠네,김선영 지음,블랙피쉬,2021.0,9788968332982,NaN,NaN,NaN,NaN,1,0,2021-03-19
3,4,예루살렘 해변,"이도 게펜 지음, 임재희 옮김",문학세계사,2021.0,9788970759906,NaN,NaN,NaN,NaN,1,0,2021-03-19
4,5,김성곤의 중국한시기행 : 장강·황하 편,김성곤 지음,김영사,2021.0,9788934990833,NaN,NaN,NaN,NaN,1,0,2021-03-19


판다스는 데이터프레임에서 기본적인 몇 가지 기술통계를 자동으로 추출해주는 describe()메서드를 제공

In [4]:
ns_book6.describe()

,번호,발행년도,도서권수,대출건수
count,379976.000000,379976.000000,379976.000000,379976.000000
mean,201726.332847,2008.516306,1.135874,11.504629
std,115836.454596,8.780529,0.483343,19.241926
min,1.000000,1947.000000,0.000000,0.000000
25%,102202.750000,2003.000000,1.000000,2.000000
50%,203179.500000,2009.000000,1.000000,6.000000
75%,301630.250000,2015.000000,1.000000,14.000000
max,401681.000000,2650.000000,40.000000,1765.000000


describe()

count 누락된 값을 제외한 데이터 개수

mean

std

min

50%

25%와 75%
순서대로 늘어놓았을 때 25% 지점과 75% 지점에 놓인 값

max

In [5]:
sum(ns_book6['도서권수'] == 0)

3206

In [6]:
ns_book7 = ns_book6[ns_book6['도서권수'] > 0]

In [7]:
ns_book7.describe(percentiles = [0.3, 0.6, 0.9])

,번호,발행년도,도서권수,대출건수
count,376770.000000,376770.000000,376770.000000,376770.000000
mean,202977.476649,2008.460076,1.145540,11.593439
std,115298.245784,8.773148,0.473853,19.279409
min,1.000000,1947.000000,1.000000,0.000000
30%,124649.700000,2004.000000,1.000000,2.000000
50%,204550.500000,2009.000000,1.000000,6.000000
60%,243537.400000,2011.000000,1.000000,8.000000
90%,361341.100000,2018.000000,2.000000,28.000000
max,401681.000000,2650.000000,40.000000,1765.000000


In [9]:
ns_book7.describe(include = 'object')

# count행은 누락된 값을 제외한 데이터 개수
# unique행은 고유한 값의 개수
# top행은 가장 많이 등장하는 값
# freq행은 top행에 등장하는 항목의 빈도수

,도서명,저자,출판사,ISBN,세트 ISBN,부가기호,권,주제분류번호,등록일자
count,376770,376770,376770,376770,55866,308252,61793,359792,376770
unique,336408,248850,21875,350810,14875,17,834,12467,4562
top,승정원일기,세종대왕기념사업회 [편],문학동네,9788937430299,9788937460005,0,1,813.6,1970-01-01
freq,250,303,4410,206,702,158235,13282,14816,28185


실행결과 '도서명'열에서 가장 많이 등장하는 데이터는 '승정원일기'로 250회

## **평균**

In [10]:
x = [10, 20, 30]
sum = 0

for i in range(3):
  sum += x[i]
print("평균: ", sum / len(x))

평균:  20.0


In [11]:
ns_book7['대출건수'].mean()

np.float64(11.593438968070707)

## **중앙값**

In [12]:
# 1 2 5(중앙값) 10 20

ns_book7['대출건수'].median()

6.0

In [14]:
# 홀수일 경우 중앙값은 가운데 숫자만 찾으면 되므로 찾기 쉬움
# 짝수일 경우는?

temp_df = pd.DataFrame([1,2,3,4])

temp_df.median()

# 가운데 두 개의 값을 평균하여 중앙값을 결정
# 1,2,3,4 에서는 2와 3을 평균한 2.5가 중앙값

,0
0,2.5


## **중복값 제거하고 중앙값 구하기**

(중요) 같은 값의 대출건수를 모두 제외하고 중앙값을 계산할 수 있다
drop_duplicate() 메서드를 사용해 '대출건수' 열에서 중복된 값을 가진 행을 제거한 후 중앙값 계산

In [16]:
ns_book7['대출건수'].drop_duplicates().median()

183.0

## **최솟값, 최댓값**

최솟값과 최댓값은 min과 max메서드로 제공

In [17]:
ns_book7['대출건수'].min()

0

In [18]:
ns_book7['대출건수'].max()

1765

## **분위수 (quartile)**

분위수는 데이터를 순서대로 늘어 놓았을 때 이를 균등한 간격으로 나누는 기준점
예를 들어 이분위수는 전체 데이터를 두 구간으로 나누기 때문에 중앙값에 해당
많이 사용하는 사분위수는 순서대로 정렬된 데이터를 네 구간으로 나눔
따라서 사분위수는 3개가 나오고 각각 25%, 50%, 75%에 해당

25%에 해당하는 값을 제 1사분위수

중앙값을 제 2사분위수

75%에 해당하는 값을 제 3사분위수


# **quantile() 메서드**

In [19]:
# 하위 25% 구간 0.25

ns_book7['대출건수'].quantile(0.25)

np.float64(2.0)

In [22]:
# 여러 개의 분위수 지정

ns_book7['대출건수'].quantile([0.25,0.5,0.75])

# 각 분위수 해당하는 값을 담은 시리즈 객체 반환

,대출건수
0.25,2.0
0.50,6.0
0.75,14.0


In [25]:
# 1,2,3,4,5가 있을 때 90% 위치에 있는 값은 얼마일가
# 시리즈 객체를 정의한 후 분위수를 구하기

pd.Series([1,2,3,4,5]).quantile(0.9)

# 4.6 출력

np.float64(4.6)

quantile() 메서드는 **interpolation매개변수**에서 중간 값을 계산하는 방법 결정
위 코드는 interpolation 매개변수를 따로 지정하지 않아서 기본값으로 계산되었음
예를 들어 4 와 5 사이에서 90%에 해당하는 분위수를 찾을 때 4의 분위수 0.75와 5의 분위수 1.0을 사용한 간단한 비례식을 세울 수 있음

이렇게 두 지점 사이에 놓인 특정 위치의 값을 구하는 방법을 보간 interpolation이라고 부름

A:a = B:b
b = (axB)/A = ((0.9-0.75)x(5-4))/(1.0-0.75) = 0.6

In [26]:
4 + (0.9-0.75) * (5-4) / (1.0-0.75)

4.6

In [27]:
pd.Series([1,2,3,4,5]).quantile(0.9, interpolation = 'midpoint')

np.float64(4.5)

interpolation 매개변수는 'linear'외에도 여러 가지 다른 보간 방식을 제공

'midpoint'는 분위수에 상관없이 무조건 두 수 사이의 중앙값을 사용

## **백분위 구하기**

앞서 분위수를 직접 지정해서 구하였는데 이번에는 반대로
1,2,3,4,5의 4의 백분위는 0.75
예로 남산 도서관 대출 데이터에서 대출건수 10이 위치한 백분위를 어떻게 찾을 수 있을까
판다스에서 제공되는 메서드는 없어 간단한 트릭으로

대출건수열의 값이 10보다 작은지 비교하여 불리언배열로 만들기

In [29]:
borrow_10_flag = ns_book7['대출건수'] < 10

In [31]:
borrow_10_flag.mean()

np.float64(0.6402712530190833)

데이터의 양이 많다면 전체 데이터를 순서대로 놓았을 때 10이 위치한 백분위와 거의 비슷하게 됨
따라서, 다음처럼 mean() 메서드를 호출하여 평균을 구하면 10보다 작은 값이 차지하는 비율을 간단히 얻을 수 있음

그럼 10에 대한 백분위는 0.65 정도가 됨
quantile()메서드에 백분위를 넣어 직접 확인해보면 기대했던 10이 출력

In [32]:
ns_book7['대출건수'].quantile(0.65)

np.float64(10.0)

## **분산 (variance)**

분산은 평균으로부터 데이터가 얼마나 퍼져있는지 나타내는 통계량
데이터가 가운데 모여있다면 분산이 작고 넓게 퍼져있다면 분산이 큼


In [33]:
ns_book7['대출건수'].var()

371.6956304306922

분산은 데이터가 평균에서 얼마나 퍼져있는지 나타내는 값이기 때문에 이 결과만 보고 값이 작은지 큰지 판단하기 어려움
분산은 제곱을 했기 때문에 데이터가 평균에서 멀어질수록 값이 급격히 커짐
또 단위가 대출건수의 제곱이 되어 해석하기 어려움
이를 보완하기 위해 분산에 제곱근을 취한 표준편차를 사용

## **표준편차 (standard deviation)구하기**

표준편차는 분산에 제곱근을 한 것으로 수식기호는 s
분산은 s^2

In [34]:
ns_book7['대출건수'].std()

19.27940949382766

도서대출 건수는 평균 11회가 조금 넘음
어떤 도서의 대출 건수를 확인했을 때 평균보다 19만큼 더 많거나 적을 수 있다는 의미
대출건수는 음수가 될 수 없기에 평균 11, 표준편차가 19면 너무 큼
아마도 평균보다 훨씬 큰 대출건수가 있기 때문에 표준편차가 큰 것으로 예상

In [35]:
# 수식으로 코드 구현

import numpy as np

diff = ns_book7['대출건수'] - ns_book7['대출건수'].mean()
np.sqrt(np.sum(diff**2) / len(ns_book7)-1)

np.float64(19.25343200307804)

## **최빈값 구하기**

기술통계의 마지막으로 최빈값
최빈값 mode 데이터에서 가장 많이 등장하는 값
describe(include = 'object')의 출력 중에 top행에 나온 값이 바로 최빈값

판다스 mode()메서드는 최빈값을 계산

In [38]:
ns_book7['도서명'].mode()

# mode는 텍스트뿐만 아니라 수치형에도 적용 가능

,도서명
0,승정원일기


In [39]:
ns_book7['발행년도'].mode()

# 남산도서관은 2012년도 도서를 가장 많이 소장
# 판다스 시리즈 객체에서 메서드 호출하여 기술통계
# 대부분 판다스 데이터프레임에서도 직접호출 가능

,발행년도
0,2012.0


## **데이터프레임에서 기술통계 구하기**

평균, 중앙값, 최솟값, 최댓값, 분위수, 분산, 표준편차를 구할 수 있는 다양한 판다스 메서드를 알아보았음
이런 메서드들은 데이터프레임에서도 사용 가능
수치형 열만 연산할 수 있기 때문에 해당 열에만 적용되도록 numeric_only매개변수를 True로 지정
ns_book7 df에서 mean()메서드를 호출하면 수치형인 네 개의 열에 대한 평균값이 계산
만약 numeric_only 매개변수를 지정하지 않으면 모든 데이터 타입 열에서 계산을 수행하기 때문에 시간이 매우 오래 걸리고 경고 발생

In [41]:
ns_book7.mean(numeric_only = True)

# '번호'열은 모두 고유한 값으로 이루어져 있기 때문에 최빈값을 찾는 것이 의미가 없음
# loc메서드로 '도서명'열부터 마지막 열까지에 대해 최빈값을 찾으면 다음과 같음

,0
번호,202977.476649
발행년도,2008.460076
도서권수,1.145540
대출건수,11.593439


In [43]:
ns_book7.loc[:, '도서명':].mode()

# 도서명열에서 가장 많이 등장하는 값은 승정원일기
# 출판사열에서 가장 많이 등장하는 값은 문학동네
# 승정원일기 =! 문학동네 출판사 도서
# mode() 메서드 결과 출력물 사이엔 서로 연관 없음 주의

,도서명,저자,출판사,발행년도,ISBN,세트 ISBN,부가기호,권,주제분류번호,도서권수,대출건수,등록일자
0,승정원일기,세종대왕기념사업회 [편],문학동네,2012.0,9788937430299,9788937460005,0,1,813.6,1,0,1970-01-01


In [44]:
ns_book7.to_csv('ns_book7.csv', index = False)

## **넘파이의 기술통계 함수**

In [45]:
# 평균

import numpy as np

np.mean(ns_book7['대출건수'])

np.float64(11.593438968070707)

In [46]:
np.average(ns_book7['대출건수'], weights = 1/ns_book7['도서권수'])

np.float64(10.543612175385386)

In [47]:
np.mean(ns_book7['대출건수'] / ns_book7['도서권수'])

np.float64(9.873029861445774)

In [48]:
ns_book7['대출건수'].sum() / ns_book7['도서권수'].sum()

np.float64(10.120503701300958)

In [49]:
# 중앙값 구하기

np.median(ns_book7['대출건수'])

np.float64(6.0)

In [50]:
# 최솟값, 최댓값 구하기

np.min(ns_book7['대출건수'])

0

In [51]:
np.max(ns_book7['대출건수'])

1765

In [52]:
# 분위수 구하기
# interpolation 매개변수가 method

np.quantile(ns_book7['대출건수'], [0.25, 0.5, 0.75])

array([ 2.,  6., 14.])

In [53]:
# 분산 구하기

np.var(ns_book7['대출건수'])

371.694643898775

In [54]:
ns_book7['대출건수'].var(ddof = 0)

371.694643898775

In [55]:
np.var(ns_book7['대출건수'], ddof = 1)

371.6956304306922

In [56]:
# 표준편차 구하기

np.std(ns_book7['대출건수'])

19.279383908693113

In [57]:
# 최빈값 구하기

values, counts = np.unique(ns_book7['도서명'], return_counts = True)
max_idx = np.argmax(counts)
values[max_idx]

'승정원일기'